clang/lib/3C/DeclRewriter.cpp

```diff
@@ -80,28 +92,71 @@ void DeclRewriter::buildItypeDecl(PVConstraint *Defn, DeclaratorDecl *Decl,
   // - Possible future change: if the internal PVConstraint is partially checked
   //   and we want to use it
   //   (https://github.com/correctcomputation/checkedc-clang/issues/704).
+  std::string Type;
   if (IsCheckedTypedef || Defn->getFV() || BaseTypeRenamed) {
     Type = Defn->mkString(Info.getConstraints(),
                           MKSTRING_OPTS(UnmaskTypedef = IsCheckedTypedef,
-                                        ForItypeBase = true));
+                                        ForItypeBase = true,
+                                        UseName = DeclName));
   } else {
     // In the remaining cases, the unchecked portion of the itype is just the
     // original type of the pointer. The first branch tries to generate the type
     // using the type and name for this specific declaration. This is important
     // because it avoids changing parameter names, particularly in cases where
     // multiple functions sharing the same name are defined in different
     // translation units.
-    if (isa_and_nonnull<ParmVarDecl>(Decl) && !Decl->getName().empty())
-      Type = qtyToStr(Decl->getType(), Decl->getNameAsString());
+    if (isa_and_nonnull<ParmVarDecl>(Decl) && !DeclName.empty())
+      Type = qtyToStr(Decl->getType(), DeclName);
     else
       Type = Defn->getOriginalTypeWithName();
   }
 
-  IType = " : itype(";
-  IType += Defn->mkString(Info.getConstraints(),
-                          MKSTRING_OPTS(EmitName = false, ForItype = true,
-                                        UnmaskTypedef = IsUncheckedTypedef));
-  IType += ")" + ABR.getBoundsString(Defn, Decl, true);
+  std::string IType = " : itype(" +
+    Defn->mkString(Info.getConstraints(),
+                   MKSTRING_OPTS(EmitName = false, ForItype = true,
+                                 UnmaskTypedef = IsUncheckedTypedef)) + ")";
+  IType += ABR.getBoundsString(Defn, Decl, true, NeedsFreshLowerBound);
+
+  std::string SDecl;
+  if (GenerateSDecls && NeedsFreshLowerBound) {
+    // For itypes, the copy of the array cannot use a checked type because we
+    // know it will be used unsafely somewhere in the body of the function.
+    // Giving it a checked type would result in Checked C type errors at the
+    // unsafe uses.
```

Comment:
I realized that if the internal PVC solved to checked, then making the supplementary decl unchecked because of `-itypes-for-extern` could actually cause a compile error.  Example:
```c
void test2(int *a, int l) {
  for(int i = 0; i < l; i++)
    a[i];
  a = a + 2;
  int *b = a;
}
```
Output of `3c -alltypes -itypes-for-extern`:
```c
void test2(int *__3c_tmp_a : itype(_Array_ptr<int>) count(l), int l) {
int *a = __3c_tmp_a;

  for(int i = 0; i < l; i++)
    a[i];
  a = a + 2;
  _Ptr<int> b = a;  // error: expression has unknown bounds, cast to ptr<T> expects source to have bounds
}
```
So I went ahead and put together a simple-minded fix in [the `range_bounds.itypes_for_extern` branch](https://github.com/correctcomputation/checkedc-clang/compare/range_bounds...range_bounds.itypes_for_extern) (independent of my other branches), and it passes the regression tests; feel free to refine it if you can.  This forced me to go ahead and factor out the duplicate code that I cited in another review comment.